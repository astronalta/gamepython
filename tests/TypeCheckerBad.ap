# aptest print_tree off
# aptest gen_code off
# aptest print_ir off

# Cycle1 and Cycle2 are type incompatible, because the blarg() function is
# not covariant (which in turn is because of the foo() function on Cycle2.
Cycle1 < Interface {
    blarg() Cycle2
}

Cycle2 < Interface {
    blarg() Cycle1
    foo() Int
}

# Type2 is a valid subtype of type1.
Type1 < Interface {
    blarg() Int {
    }
}

Type2 < Interface {

}

Object1 < Object {
    blarg() Int

    @index(x Type1, y Type1) {
    }
}

# Fully-qualified names are not allowed (now handled by parser).
#Scoped::Illegal < Interface {
#    blarg() Int
#}

Value1 < Value {
}

free_function() {
}

TypeCheckerBad < Object {

    blarg(blarg Int) {}
    void() {}
    test_cycle1(x Cycle1) {}
    test_cycle2(x Cycle2) {}
    test_type1(x Type1) {}
    test_type2(x Type2) {}
    test_object1(x Object1) {}

    main() {
        a Int = "Hello, world" # Incompatible type
        b = "Hello, world" 
        b String # Variable redefinition
        b Int # Variable redefinition, different type
        
        c Foo # Defuned class 
        c.whatever() 
        
        d = TypeCheckerBad().blarg("Hello, world") # Assign void
        d String # Redefine variable
        d = 1 # OK, because d has no type

        cycle1 Cycle1
        cycle2 Cycle2
        type1 Type1
        type2 Type2
        object1 Object1

        cycle1 = cycle2 # Invalid type
        type1 = type2 # Invalid type
        object1 = type1 # Invalid type
        type1 = object1 # OK
        
        e = TypeCheckerBad()

        e.missing() # Undefined function
        e.test_cycle1(cycle2) # Invalid arg
        e.test_cycle1(cycle1) # OK
        e.test_cycle2(cycle1) # Invalid arg
        e.test_type1(type2) # Invalid arg
        e.test_type1(type1) # OK
        e.test_type2(type2) # OK
        e.test_type2(type1) # OK

        Cycle1() # Constructor on an interface

        e.test_cycle1(cycle1, "Hello", 1) # Too many args
        e.test_cycle1() # Not enough args

        object1[1, 2] # Bad arguments

        f Value1
        if f {} # Using struct as a bool
        if f and true {} # Using struct as a bool
        if not f {} # Using struct as a bool
    }
}
