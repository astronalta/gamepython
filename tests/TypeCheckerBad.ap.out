18:17: Interface method 'blarg' has a block
27:5: Method 'blarg' has no block
55:17: Expression does not conform to type 'Int'
57:9: Duplicate definition of variable 'b'
58:9: Duplicate definition of variable 'b'
60:11: Undefined type 'Foo'
63:36: Argument does not conform to type 'Int'
63:13: Void value assigned to variable 'd'
64:9: Duplicate definition of variable 'd'
73:18: Expression does not conform to type 'Cycle1' in assignment of 'cycle1'
74:17: Expression does not conform to type 'Type1' in assignment of 'type1'
75:19: Expression does not conform to type 'Object1' in assignment of 'object1'
80:9: Undeclared function 'missing'
81:23: Argument does not conform to type 'Cycle1'
83:23: Argument does not conform to type 'Cycle2'
84:22: Argument does not conform to type 'Type1'
89:9: Constructor called for interface type 'Cycle1'
91:9: Too many arguments to function 'test_cycle1'
92:9: Not enough arguments to function 'test_cycle1'
94:17: Argument does not conform to type 'Type1'
94:20: Argument does not conform to type 'Type1'
97:12: Value types cannot be converted to 'Bool'
98:12: Value types cannot be converted to 'Bool'
99:12: Value types cannot be converted to 'Bool'
